From 4c83d06bfdcc37f9b8de824671235238a9e48cec Mon Sep 17 00:00:00 2001
From: Sayanta Pattanayak <sayanta.pattanayak@arm.com>
Date: Mon, 10 Jan 2022 18:35:58 +0530
Subject: [PATCH] module/cxl: add support for CXL

Pcie_enumeration module invokes the CXL module API for discovering PCIe
device with extended CXL capability. Internally CXL module also looks
for CXL Device with DOE capability. Once DOE capability is found, then
execute DOE operation to fetch CDAT structures, which carry
information about CXL Device Memory range, type of memory etc. Based on
CDAT structure information and CXL device memory range supported, CMN700
runtime API will be invoked for mapping designated Host address space to
CXL device memory area.

Signed-off-by: Sayanta Pattanayak <sayanta.pattanayak@arm.com>
Change-Id: I77ae2024c1f1dcdc79cb06f2bc88d7de03a3e7ba

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 module/cxl/CMakeLists.txt                     |  20 +
 module/cxl/Module.cmake                       |   9 +
 module/cxl/include/mod_cxl.h                  | 243 +++++++++++
 module/cxl/src/Makefile                       |  11 +
 module/cxl/src/mod_cxl.c                      | 406 ++++++++++++++++++
 .../neoverse-rd/rdn2/scp_ramfw/Firmware.cmake |   5 +-
 6 files changed, 693 insertions(+), 1 deletion(-)
 create mode 100644 module/cxl/CMakeLists.txt
 create mode 100644 module/cxl/Module.cmake
 create mode 100644 module/cxl/include/mod_cxl.h
 create mode 100644 module/cxl/src/Makefile
 create mode 100644 module/cxl/src/mod_cxl.c

diff --git a/module/cxl/CMakeLists.txt b/module/cxl/CMakeLists.txt
new file mode 100644
index 000000000000..fc712a2ebc07
--- /dev/null
+++ b/module/cxl/CMakeLists.txt
@@ -0,0 +1,20 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(
+    ${SCP_MODULE_TARGET}
+    PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+
+target_sources(
+    ${SCP_MODULE_TARGET}
+    PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_cxl.c")
+
+target_link_libraries (
+    ${SCP_MODULE_TARGET} PRIVATE module-apremap module-sds
+                                 module-pcie_discovery module-cmn700)
diff --git a/module/cxl/Module.cmake b/module/cxl/Module.cmake
new file mode 100644
index 000000000000..99c426e678e5
--- /dev/null
+++ b/module/cxl/Module.cmake
@@ -0,0 +1,9 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "cxl")
+set(SCP_MODULE_TARGET "module-cxl")
diff --git a/module/cxl/include/mod_cxl.h b/module/cxl/include/mod_cxl.h
new file mode 100644
index 000000000000..8ae399dcab1f
--- /dev/null
+++ b/module/cxl/include/mod_cxl.h
@@ -0,0 +1,243 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     CXL module header file.
+ */
+
+#ifndef MOD_CXL_H
+#define MOD_CXL_H
+
+#include <fwk_macros.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#define WORD_MASK     UINT64_C(0xFFFFFFFF)
+#define WORD_SHIFT    UINT64_C(32)
+
+/* Extended Capabilities (PCI-X 2.0 and Express) */
+#define PCI_EXT_CAP_ID(header)         ((header) & 0x0000ffff)
+#define PCI_EXT_CAP_VER(header)        (((header) >> 16) & 0xf)
+#define PCI_EXT_CAP_NEXT(header)       (((header) >> 20) & 0xfff)
+#define PCI_CFG_SPACE_SIZE             256
+#define PCI_CFG_SPACE_EXP_SIZE         4096
+
+/* PCIe DVSEC for CXL Device */
+#define DVSEC_CXL_HEADER1              0x4
+#define DVSEC_VENDOR_ID_MASK           0xFFFF
+#define DVSEC_CXL_VENDOR_ID            0x1E98
+#define DVSEC_REVISION_MASK            0xF
+#define DVSEC_REVISION_BIT_POS         16
+#define DVSEC_CXL_REVISION             0x1
+
+#define DVSEC_CXL_HEADER2              0x8
+#define DVSEC_CXL_DEVICE_ID_MASK       0xFFFF
+#define DVSEC_CXL_DEVICE_ID            0x0
+
+#define DVSEC_CXL_CTRL_STATUS          0xC
+#define DVSEC_CXL_CTRL_STATUS_2        0x10
+#define DVSEC_CXL_LOCK_CAP2            0x14
+#define DVSEC_CXL_RANGE_1_SIZE_HIGH    0x18
+#define DVSEC_CXL_RANGE_1_SIZE_LOW     0x1C
+#define DVSEC_CXL_RANGE_1_BASE_HIGH    0x20
+#define DVSEC_CXL_RANGE_1_BASE_LOW     0x24
+#define DVSEC_CXL_RANGE_2_SIZE_HIGH    0x28
+#define DVSEC_CXL_RANGE_2_SIZE_LOW     0x2C
+#define DVSEC_CXL_RANGE_2_BASE_HIGH    0x30
+#define DVSEC_CXL_RANGE_2_BASE_LOW     0x34
+
+#define CXL_DEVICE_CAP_ARRAY_REG1      0x0
+#define CXL_DEVICE_CAP_ARRAY_REG2      0x4
+#define CXL_DEVICE_CAP1_HEADER_REG1    0x10
+#define CXL_DEVICE_CAP1_HEADER_REG2    0x14
+#define CXL_DEVICE_CAP1_HEADER_REG3    0x18
+#define CXL_DEVICE_CAP2_HEADER_REG1    0x20
+#define CXL_DEVICE_CAP2_HEADER_REG2    0x24
+#define CXL_DEVICE_CAP2_HEADER_REG3    0x28
+
+#define MAILBOX_CAP_REG                0x0
+#define MAILBOX_CONTROL_REG            0x4
+#define MAILBOX_COMMAND_REG1           0x8
+#define MAILBOX_COMMAND_REG2           0xC
+#define MAILBOX_STATUS_REG1            0x10
+#define MAILBOX_STATUS_REG2            0x14
+#define MAILBOX_BG_CMD_STATUS_REG      0x18
+#define MAILBOX_COMMAND_PAYLOAD_REG    0x20
+
+#define CXLDEV_MBOX_CTRL_DOORBELL      BIT(0)
+
+#define DOE_DATA_OBJ_HEADER_1          0x0
+#define DOE_DATA_OBJ_HEADER_2          0x4
+
+#define DOE_CAPABILITIES_REG           0x4
+#define DOE_CONTROL_REG                0x8
+#define DOE_STATUS_REG                 0xC
+#define DOE_WRITE_DATA_MAILBOX_REG     0x10
+#define DOE_READ_DATA_MAILBOX_REG      0x14
+
+#define PCIE_DVSEC_ID                  0x23
+#define PCIE_DOE_ID                    0x2E
+
+#define DOE_DATA_OBJECT_LENGTH         0x0003ffff
+#define CXL_DOE_TABLE_ENTRY_HANDLE     0xffff0000
+#define CXL_DOE_TABLE_ENTRY_HANDLE_LAST 0xffff
+
+#define   DOE_STAT_DOE_BUSY            UINT32_C(0x1)
+#define   DOE_STAT_DATA_OBJ_READY      (UINT32_C(0x1) << 31)
+#define   DOE_CTRL_DOE_GO              (UINT32_C(0x1) << 31)
+
+/*!
+ * \brief Data object header
+ *
+ * \details Data Object Exchange(DOE) Header1 and Header2 put together.
+ *      Reference taken from PCI-SIG ECN and
+ *      CXL Specification (Revision 3.0, Version 0.7).
+ */
+struct DOEHeader {
+    uint16_t vendor_id;
+    uint8_t data_obj_type;
+    uint8_t reserved;
+    uint32_t length;
+};
+
+#define DOE_DATA_OBJ_TYPE_COMPLIANCE   0x0
+#define DOE_DATA_OBJ_TYPE_CDAT         0x2
+
+/*!
+ * \brief DOE read request data
+ *
+ * \details DOE read request data structure. For CXL, DOE requests are made
+ *      to read CDAT tables.
+ *      Reference taken from CXL Specification (Revision 3.0, Version 0.7).
+ */
+struct cxl_cdat_read_entry_req {
+    struct DOEHeader header;
+    uint8_t req_code;
+    uint8_t table_type;
+    uint16_t entry_handle;
+};
+
+#define CXL_CDAT_DOE_ENTRYHANDLE_LAST_ENTRY  0xFFFF
+/* Size in DW(4 Bytes) */
+#define CDAT_READ_ENTRY_REQ_SIZE    3
+
+/*!
+ * \brief DOE read response data
+ *
+ * \details DOE read response data structure. For CXL, DOE responses carry
+ *      information about CDAT tables.
+ *      Reference taken from CXL Specification (Revision 3.0, Version 0.7).
+ */
+struct cxl_cdat_read_entry_resp {
+    struct DOEHeader header;
+    uint8_t rsp_code;
+    uint8_t table_type;
+    uint16_t entry_handle;
+    uint32_t cdat_table[32];
+};
+
+/* Size in DW(4 Bytes) */
+#define CDAT_READ_ENTRY_RESP_SIZE    3
+
+/*!
+ * \brief Coherent Device Attribute Table(CDAT) Header
+ *
+ * \details CDAT header, which is followed by variable number of CDAT structures.
+ *      Reference taken from CDAT Specification (Revision 1.02).
+ */
+struct cdat_table_header {
+    uint32_t length;
+    uint8_t revision;
+    uint8_t checksum;
+    uint8_t reserved[6];
+    uint32_t sequence;
+};
+
+/* Size in DW(4 Bytes) */
+#define CDAT_TABLE_HEADER_SIZE    4
+
+/*!
+ * \brief Device Scoped Memory Affinity Structure (DSMAS)
+ *
+ * \details DSMAS returns Device Physical Address(DPA) range and it's attributes.
+ *      Reference taken from CDAT Specification (Revision 1.02).
+ */
+struct cdat_dsmas {
+    uint8_t type;
+    uint8_t reserved_1;
+    uint16_t length;
+    uint8_t DSMADhandle;
+    uint8_t flags;
+    uint16_t reserved_2;
+    uint64_t DPA_base;
+    uint64_t DPA_length;
+};
+
+/* Size in DW(4 Bytes) */
+#define CDAT_STRUCTURE_DSMAS_SIZE   6
+
+/* Size in DW(4 Bytes) */
+#define CDAT_STRUCTURE_DSLBIS_SIZE   6
+
+struct cxl_cdat_dsmas_doe_resp {
+    struct cxl_cdat_read_entry_resp resp;
+    struct cdat_table_header header;
+    struct cdat_dsmas       dsmas;
+};
+
+/*!
+ * \brief Type of CDAT structures
+ */
+enum cdat_structure_type {
+    CDAT_STRUCTURE_DSMAS = 0,
+    CDAT_STRUCTURE_DSLBIS,
+    CDAT_STRUCTURE_DSMSCIS,
+    CDAT_STRUCTURE_DSIS,
+    CDAT_STRUCTURE_DSEMTS,
+    CDAT_STRUCTURE_SSLBIS,
+    CDAT_STRUCTURE_COUNT
+};
+
+struct cxl_mem_region {
+    uint64_t cxl_mem_base;
+    uint64_t free_cxl_mem_base;
+    uint64_t cxl_mem_size;
+};
+
+/*!
+ * \brief Module interface.
+ */
+struct mod_cxl_api {
+    /*!
+     * \brief Find out whether a PCIe device supports CXL extended capability.
+     *
+     * \param Ecam base address of PCIe device, which will be  checked for
+     *      device capability.
+     */
+    void (*find_cxl_cap)(uint64_t config_base);
+
+    /*!
+     * \brief Call CMN api to map host address space to CXL mem region.
+     */
+    void (*map_cxl_mem_region)(void);
+};
+
+/*!
+ * \brief Remote memory details
+ *
+ * \details It will hold combined details of host address space and
+ *      device address space.
+ */
+struct remote_memory_config {
+    uint32_t  addressbaselow;
+    uint32_t  addressbasehigh;
+    uint32_t  lengthlow;
+    uint32_t  lengthhigh;
+};
+
+void update_cxl_mem_region(uint64_t mem_base, uint64_t size);
+#endif /* MOD_CXL_H */
diff --git a/module/cxl/src/Makefile b/module/cxl/src/Makefile
new file mode 100644
index 000000000000..f5c65bf4d324
--- /dev/null
+++ b/module/cxl/src/Makefile
@@ -0,0 +1,11 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+BS_LIB_NAME := CXL
+BS_LIB_SOURCES = mod_cxl.c
+
+include $(BS_DIR)/lib.mk
diff --git a/module/cxl/src/mod_cxl.c b/module/cxl/src/mod_cxl.c
new file mode 100644
index 000000000000..6e6e47dcd257
--- /dev/null
+++ b/module/cxl/src/mod_cxl.c
@@ -0,0 +1,406 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     CXL module for discovering CXL capable PCIe device.
+ *     Execute DOE operation.
+ */
+
+#include <interface_cmn.h>
+
+#include <mod_apremap.h>
+#include <mod_cxl.h>
+#include <mod_pcie_discovery.h>
+
+#include <fwk_assert.h>
+#include <fwk_event.h>
+#include <fwk_id.h>
+#include <fwk_list.h>
+#include <fwk_log.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+#include <fwk_notification.h>
+#include <fwk_status.h>
+
+#define MOD_NAME "[PCIE_CXL] "
+
+#define MAX_REMOTE_MEM  5
+
+/* AP remap Adress read write APIs */
+static struct interface_address_remapper_rw_api *remapper_rw_api;
+
+/* APIs to configure CMN700 moudle */
+static struct interface_cmn_memmap_rnsam_api *cmn_api;
+
+/* AP remap Address translation enable/disable APIs */
+struct mod_apremap_cmn_atrans_api *apremap_cmn_atrans_api;
+
+static struct remote_memory_config remote_mem[MAX_REMOTE_MEM];
+static uint32_t next_remote_mem;
+static struct cxl_mem_region host_cxl_mem;
+static uint64_t remote_mem_size;
+
+int is_doe_busy(uint64_t doe_stat_reg,
+                struct pcie_discovery_rw_api rw_api)
+{
+    uint32_t doe_stat_val;
+
+    doe_stat_val = rw_api.read32(doe_stat_reg);
+    if (doe_stat_val & DOE_STAT_DOE_BUSY)
+        return FWK_E_BUSY;
+
+    return FWK_SUCCESS;
+}
+
+static void update_cdat_data(uint32_t *doe_resp_cdat_data, uint16_t length)
+{
+    uint32_t cur_index;
+    struct cdat_dsmas *dsmas = NULL;
+    uint64_t dev_phys_addr;
+
+    /* skipping the CDAT header. */
+    cur_index = 4;
+    while (cur_index < length) {
+        switch ((doe_resp_cdat_data[cur_index] & 0xff)) {
+        case CDAT_STRUCTURE_DSMAS:
+            dsmas = (struct cdat_dsmas *)(&(doe_resp_cdat_data[cur_index]));
+            dev_phys_addr = host_cxl_mem.free_cxl_mem_base + dsmas->DPA_base;
+            remote_mem[next_remote_mem].addressbaselow =
+                dev_phys_addr & WORD_MASK;
+            remote_mem[next_remote_mem].addressbasehigh =
+                (dev_phys_addr >> WORD_SHIFT) & WORD_MASK;
+            remote_mem[next_remote_mem].lengthlow =
+                dsmas->DPA_length & WORD_MASK;
+            remote_mem[next_remote_mem].lengthhigh =
+                dsmas->DPA_length >> WORD_SHIFT;
+            host_cxl_mem.free_cxl_mem_base += dsmas->DPA_length;
+            next_remote_mem++;
+            cur_index += CDAT_STRUCTURE_DSMAS_SIZE;
+            remote_mem_size += dsmas->DPA_length;
+            break;
+        case CDAT_STRUCTURE_DSLBIS:
+            cur_index += CDAT_STRUCTURE_DSLBIS_SIZE;
+            break;
+        default:
+            break;
+        }
+    }
+
+    return;
+}
+
+/* Receive DOE response data a DWORD at a time. */
+
+static uint16_t doe_receive_response(uint64_t doe_base,
+                             struct pcie_discovery_rw_api rw_api)
+{
+    uint32_t doe_resp_cdat_data[15] = {};
+    uint32_t cur_index = 0;
+    uint32_t data;
+    uint32_t val;
+    uint64_t reg;
+    uint16_t length;
+    uint16_t entry_handle = 0;
+
+    reg = doe_base + DOE_STATUS_REG;
+    val = rw_api.read32(reg);
+
+    if (val & DOE_STAT_DATA_OBJ_READY) {
+        cur_index = 0;
+        reg = doe_base + DOE_READ_DATA_MAILBOX_REG;
+
+        /* Read the DOE header. */
+        data = rw_api.read32(reg);
+        rw_api.write32(reg, 0x1);
+
+        /* Read the DOE Header 2 for data length. */
+        data = rw_api.read32(reg);
+        length = data & DOE_DATA_OBJECT_LENGTH;
+        rw_api.write32(reg, 0x1);
+
+        if (length < 2) {
+            FWK_LOG_WARN(MOD_NAME " DOE data read error \n");
+            return 0;
+        }
+
+        /* Read DOE read entry response header. */
+        data = rw_api.read32(reg);
+        entry_handle = ((data & CXL_DOE_TABLE_ENTRY_HANDLE) >> 16);
+        rw_api.write32(reg, 0x1);
+
+        length -= 3;
+        while (cur_index < length) {
+            doe_resp_cdat_data[cur_index] = rw_api.read32(reg);
+            cur_index++;
+            rw_api.write32(reg, 0x1);
+        }
+
+        update_cdat_data(doe_resp_cdat_data, length);
+    }
+
+    return entry_handle;
+}
+
+/*
+ * Execute DOE command to fetch CDAT structures.
+ *
+ * The steps include -
+ * 1. System firmware checks that the DOE Busy bit is Clear.
+ * 2. System firmware writes entire data object a DWORD at a time via
+ *    DOE Write Data Mailbox register.
+ * 3. System firmware writes 1b to the DOE Go bit.
+ * 4. The DOE instance consumes the DOE request from the DOE mailbox.
+ * 5. The DOE instance generates a DOE Response and Sets Data Object Ready bit.
+ * 6. System firmware polls the Data Object Ready bit and, provided it is Set,
+ *    reads data from the DOE Read Data Mailbox and writes to the DOE Read Data
+ *    Mailbox to indicate a successful read, a DWORD at a time until the entire
+ *    DOE Response is read.
+ *
+ */
+static void send_doe_command(uint64_t doe_base,
+                             struct pcie_discovery_rw_api rw_api)
+{
+    uint32_t val;
+    uint64_t reg;
+    uint32_t cur_index = 0;
+    struct cxl_cdat_read_entry_req cxl_doe_req;
+
+    /* CDAT DOE Request header & Read entry request object */
+    cxl_doe_req.header.vendor_id = DVSEC_CXL_VENDOR_ID;
+    cxl_doe_req.header.data_obj_type = DOE_DATA_OBJ_TYPE_CDAT;
+    cxl_doe_req.header.length = CDAT_READ_ENTRY_REQ_SIZE;
+
+    /* 0 indicates this a request to read */
+    cxl_doe_req.req_code = 0;
+
+    /* 0 indicates table type as CDAT */
+    cxl_doe_req.table_type = 0;
+
+    /* 0 represents very first entry in the table */
+    cxl_doe_req.entry_handle = 0;
+
+    reg = doe_base + DOE_WRITE_DATA_MAILBOX_REG;
+
+    do {
+        if (is_doe_busy(doe_base + DOE_STATUS_REG, rw_api)) {
+            FWK_LOG_WARN(MOD_NAME " DOE busy \n");
+            return;
+        }
+
+        while (cur_index < CDAT_READ_ENTRY_REQ_SIZE) {
+            rw_api.write32(
+                reg,
+                *((uint32_t *)(&cxl_doe_req) + cur_index));
+            cur_index++;
+        }
+
+        reg = doe_base + DOE_CONTROL_REG;
+        val = rw_api.read32(reg);
+        val |= DOE_CTRL_DOE_GO;
+        rw_api.write32(reg, val);
+
+        cxl_doe_req.entry_handle = doe_receive_response(doe_base, rw_api);
+
+    } while ( cxl_doe_req.entry_handle < CXL_DOE_TABLE_ENTRY_HANDLE_LAST);
+
+    return;
+}
+
+/*
+ * Find out whether the PCIe device supports CXL extended capability.
+ */
+static uint16_t find_pcie_ext_cap(uint64_t config_base,
+                                  uint16_t ext_cap_off, uint16_t cap_id,
+                                  struct pcie_discovery_rw_api rw_api)
+{
+    uint64_t cap_offset;
+    uint32_t cap_val;
+
+    cap_offset = config_base + ext_cap_off;
+
+    do {
+        cap_val =  rw_api.read32(cap_offset);
+        if ((PCI_EXT_CAP_ID(cap_val) == cap_id)) {
+            cap_val = rw_api.read32(cap_offset +
+                DVSEC_CXL_HEADER1);
+            if ((cap_val & 0xffff) == DVSEC_CXL_VENDOR_ID)
+                return ext_cap_off;
+        }
+
+        ext_cap_off = PCI_EXT_CAP_NEXT(cap_val);
+        cap_offset = config_base + ext_cap_off;
+    } while (ext_cap_off);
+
+   return 0;
+}
+
+/*
+ * Find out whether the PCIe device supports DOE capability.
+ */
+static uint16_t find_pcie_doe_cap(uint64_t config_base,
+                                  uint16_t ext_cap_off, uint16_t cap_id,
+                                  struct pcie_discovery_rw_api rw_api)
+{
+    uint64_t cap_offset;
+    uint32_t cap_val;
+
+    cap_offset = config_base + ext_cap_off;
+
+    do {
+        cap_val =  rw_api.read32(cap_offset);
+        if (PCI_EXT_CAP_ID(cap_val) == cap_id)
+            return ext_cap_off;
+
+        ext_cap_off = PCI_EXT_CAP_NEXT(cap_val);
+        cap_offset = config_base + ext_cap_off;
+    } while (ext_cap_off);
+
+   return 0;
+}
+
+/*
+ * This function gets the details about the Host physical address space, which
+ * is reserved and configured for  CXL Memory.
+ */
+void update_cxl_mem_region(uint64_t mem_base, uint64_t size)
+{
+    host_cxl_mem.cxl_mem_base = mem_base;
+    host_cxl_mem.free_cxl_mem_base = mem_base;
+    host_cxl_mem.cxl_mem_size = size;
+
+}
+
+void map_cxl_mem_region(void)
+{
+   int status;
+
+   apremap_cmn_atrans_api->enable();
+   status = cmn_api->map_ccg_for_cxl_mem (remote_mem_size);
+   if (status != FWK_SUCCESS) {
+       FWK_LOG_ERR(MOD_NAME "CXL Mem region mapping failed, Status %d", status);
+       return;
+   }
+   apremap_cmn_atrans_api->disable();
+}
+
+void find_cxl_capability(uint64_t config_base)
+{
+    uint32_t offset = 0;
+    uint64_t cap_offset;
+    uint32_t vendor_id;
+    uint32_t cxl_id;
+    uint64_t cap_val;
+    struct pcie_discovery_rw_api rw_api;
+
+    rw_api.read8 = remapper_rw_api->read8;
+    rw_api.write8 = remapper_rw_api->write8;
+    rw_api.read32 = remapper_rw_api->read32;
+    rw_api.write32 = remapper_rw_api->write32;
+
+    offset = find_pcie_ext_cap(config_base,
+                               PCI_CFG_SPACE_SIZE,
+                               PCIE_DVSEC_ID,
+                               rw_api);
+    if (!offset) {
+        return;
+    }
+
+    cap_offset = config_base + offset + DVSEC_CXL_HEADER1;
+    cap_val =  rw_api.read32(cap_offset);
+    vendor_id = cap_val & DVSEC_VENDOR_ID_MASK;
+
+    cap_offset = config_base + offset + DVSEC_CXL_HEADER2;
+    cap_val =  rw_api.read32(cap_offset);
+    cxl_id = cap_val & DVSEC_CXL_DEVICE_ID_MASK;
+
+    if ((vendor_id != DVSEC_CXL_VENDOR_ID) || (cxl_id != DVSEC_CXL_DEVICE_ID)) {
+        FWK_LOG_WARN(MOD_NAME "No CXL Device found");
+        return;
+    }
+
+    offset = find_pcie_doe_cap(config_base,
+                               PCI_CFG_SPACE_SIZE,
+                               PCIE_DOE_ID,
+                               rw_api);
+    if (!offset) {
+        FWK_LOG_WARN(MOD_NAME "DOE Capability Not Found \n");
+        return;
+    }
+
+    send_doe_command(config_base + offset, rw_api);
+
+    return;
+}
+
+static int cxl_bind(fwk_id_t id, unsigned int round)
+{
+    int status;
+
+    status = fwk_module_bind(
+        FWK_ID_MODULE(FWK_MODULE_IDX_APREMAP),
+        FWK_ID_API(FWK_MODULE_IDX_APREMAP, MOD_APREMAP_API_IDX_AP_MEM_RW),
+        &remapper_rw_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    status = fwk_module_bind(
+                 FWK_ID_MODULE(FWK_MODULE_IDX_CMN700),
+                 FWK_ID_API(FWK_MODULE_IDX_CMN700, 0),
+                 &cmn_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    status = fwk_module_bind(
+                 FWK_ID_MODULE(FWK_MODULE_IDX_APREMAP),
+                 FWK_ID_API(FWK_MODULE_IDX_APREMAP,
+                 MOD_APREMAP_API_IDX_CMN_ATRANS),
+                 &apremap_cmn_atrans_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int cxl_init(fwk_id_t module_id,
+                    unsigned int element_count,
+                    const void *unused)
+{
+    return FWK_SUCCESS;
+}
+
+static struct mod_cxl_api cxl_api = {
+    .find_cxl_cap = find_cxl_capability,
+    .map_cxl_mem_region = map_cxl_mem_region,
+};
+
+static int cxl_bind_request(fwk_id_t requester_id,
+                            fwk_id_t id,
+                            fwk_id_t api_id,
+                            const void **api)
+{
+
+    if (!fwk_module_is_valid_module_id(requester_id)) {
+        return FWK_E_ACCESS;
+    }
+
+    *api = &cxl_api;
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module_config config_cxl = { 0 };
+
+const struct fwk_module module_cxl = {
+    .type = FWK_MODULE_TYPE_SERVICE,
+    .api_count = 1,
+    .event_count = 0,
+    .init = cxl_init,
+    .process_bind_request = cxl_bind_request,
+    .bind = cxl_bind,
+};
diff --git a/product/neoverse-rd/rdn2/scp_ramfw/Firmware.cmake b/product/neoverse-rd/rdn2/scp_ramfw/Firmware.cmake
index da6dafe70376..aa90c6478a26 100644
--- a/product/neoverse-rd/rdn2/scp_ramfw/Firmware.cmake
+++ b/product/neoverse-rd/rdn2/scp_ramfw/Firmware.cmake
@@ -45,8 +45,10 @@ list(PREPEND SCP_MODULE_PATHS
      "${CMAKE_CURRENT_LIST_DIR}/../module/pcie_integ_ctrl")
 list(PREPEND SCP_MODULE_PATHS
      "${CMAKE_CURRENT_LIST_DIR}/../../module/pcie_setup")
- list(PREPEND SCP_MODULE_PATHS
+list(PREPEND SCP_MODULE_PATHS
      "${CMAKE_SOURCE_DIR}/module/pcie_discovery")
+list(PREPEND SCP_MODULE_PATHS
+     "${CMAKE_SOURCE_DIR}/module/cxl")
 
 # The order of the modules in the following list is the order in which the
 # modules are initialized, bound, started during the pre-runtime phase.
@@ -78,6 +80,7 @@ list(APPEND SCP_MODULES "scmi-system-power")
 list(APPEND SCP_MODULES "pcie_discovery")
 list(APPEND SCP_MODULES "pcie-setup")
 list(APPEND SCP_MODULES "pcie_integ_ctrl")
+list(APPEND SCP_MODULES "cxl")
 list(APPEND SCP_MODULES "platform-system")
 list(APPEND SCP_MODULES "sp805")
 list(APPEND SCP_MODULES "psu")
